### DISCUSSION TITLE
Porting applications *.desktop file to freedesktop spec

### DISCUSSION THREAD

**Discussion by stefonarch** [lxqt/lxqt#2448] [General]
>The name of the desktop entry should follow the "reverse DNS" convention: it should start with a reversed DNS domain name controlled by the author of the application, in lower case. The domain name should be followed by the name of the application, which is conventionally written with words run together and initial capital letters (CamelCase). For example, if the owner of example.org writes "Foo Viewer", they might choose the name org.example.FooViewer, resulting in a file named org.example.FooViewer.desktop.
>
>Well-known names containing the dash are allowed but not recommended, because the dash is not allowed in some related uses of reversed DNS names, such as D-Bus object paths and interface names, and Flatpak app ID

https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s02.html

It's some time now in `/usr/share/applications/` this format shows up, examples:

```
org.kde.okular.desktop
org.kde.partitionmanager.desktop
org.qbittorrent.qBittorrent.desktop
org.qt-project.qtcreator.desktop
org.strawberrymusicplayer.strawberry.desktop
org.gnome.FontManager.desktop
org.gnome.FontViewer.desktop
io.elementary.granite.demo.desktop
io.element.Element.desktop
```

I've started already some PRs, but I'm not sure about the format we should use. Personally I like the capitalized form for the app as in it's name and keeping the dash in the application name like

 ```
org.lxqt_project.QTerminal.desktop 
org.lxqt_project.QTerminal-drop.desktop
org.lxqt_project.LXimage-qt.desktop
 org.lxqt_project.ObConf-Qt.desktop

```
Other options are
1. all lowercase (like kde does) and
2. all hyphens or
3. all dashes (as qt-project does, ignoring the recommendation)

Thoughts?

**Comment by isf63** [comment]
From the spec:

> ... should start with a reversed DNS domain name controlled by the author of the application ...
> 
> The domain name should be followed by the name of the application, which is conventionally written with words run together and initial capital letters (CamelCase).

So `org.lxqt-project.ObConf-Qt.desktop` would follow this, disregarding the CamelCase convention. Many standalone applications ignore this "reverse DNS" (looking at my /usr/share/applications/), but it might be good for desktop environments.

**Comment by redtide** [comment]
My thoughts about this after some more reading:
when something is recommended in a spec it's only confusing, if it's a spec, it should define one way or another (mandatory), not both, so I see this way:

- Lower case domain names.
- Application names "CamelCase"? To be picky, it should be `PascalCase` if anything (like function names in wxWidgets and class names in Qt), otherwise `camelCase` (like class names in wxWidgets and function names in Qt 😄) with no initial capitalization. This is not much clear.
- Underscores everywhere replacing dashes, with names starting with a digit underscore prefixed.
- Dash are tolerated for backward compatibility so it's supported but SHOULD not be used for new names.

So if "CamelCase" should not contain underscores, nor dashes IIUC, (e.g.: ObConfQt and LxImage).

Basically all of this comes from recent DBus naming convention rules, described in the "[D-Bus well-known name](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names)" link to the DBus spec in the above link:

> Each element must only contain the ASCII characters "[A-Z][a-z][0-9]_-", with "-" discouraged in new bus names. Only elements that are part of a unique connection name may begin with a digit, elements in other bus names must not begin with a digit.

**Comment by isf63** [comment]
While I don't prefer underscores over hyphens, if these files are really used by programs (D-Bus, Flatpak) with that being disallowed, it's tolerable.

Care should be taken to preserve the programs' title capitalizations - i.e. PCManFM-Qt, LXImage-Qt, QTerminal, ...

**Reply by yan12125** [reply]
> if these files are really used by programs (D-Bus, Flatpak) with that being disallowed, it's tolerable.

It's for `DBusActivatable` feature when such a naming scheme is first proposed [1]. Not sure about Flatpak.

[1] https://gitlab.freedesktop.org/xdg/xdg-specs/-/commit/6650489387dbc2c03282831dcddfd2af8934d513

**Comment by yan12125** [comment]
Some comments at the first glance:

* How about `org.lxqt` instead of `org.lxqt_project`? We already have some `org.lxqt` names:
```
$ pacman -Fx 'org\.lxqt'
extra/liblxqt 1.3.0-1
    usr/share/polkit-1/actions/org.lxqt.backlight.pkexec.policy
extra/lxqt-admin 1.3.0-1 (lxqt)
    usr/share/polkit-1/actions/org.lxqt.lxqt-admin-user.policy
```
I know lxqt.org has been dead for a long time. That's just an identifier, anyway.

* This change may be backward incompatible. If a user overrides some `.desktop` files in `~/.config/autostart` or `~/.local/share/applications`, they need to update those overrides. I'm fine with such a backward incompatibility if release notes mention that.

**Reply by tsujan** [reply]
> This change may be backward incompatible. If a user overrides some .desktop files in ~/.config/autostart or...

There are more examples of backward incompatibility. For example, it interferes with the user's `*-mimeapps.list` files. IMO, that's not OK. If it was up to me (which isn't, of course), I wouldn't annoy users to this extent. Is this rule so essential to justify that?

So, I think it's better to limit it to apps without mime type association (like Qps, lxqt-config,...).

**Reply by stefonarch** [reply]
I agree with file associations to look closely at, maybe not porting lximage-qt and lxqt-archiver without extent testing. Here it looks like they are preserved.

Using `lxqt.org` would be nice but is against the spec as it is not under our control, but why not?

**Reply by stefonarch** [reply]
Hm, org.lxqt is used already in dbus, so that's a point more for using this.

![screen_area_dom_16:45:30_](https://github.com/lxqt/lxqt/assets/10681413/0b3d7f80-03a9-4df5-89ef-23aa73df5010)

**Reply by yan12125** [reply]
> For example, it interferes with the user's `*-mimeapps.list` files.

Good reminder about file associations! There is only one more case (pcmanfm-qt.desktop): https://github.com/search?q=org%3Alxqt+MimeType+language%3Adesktop&type=code&l=desktop. I don't count lxqt-organizer as it's unmaintained and the MimeType declaration is suspicious.

---

> Using `lxqt.org` would be nice but is against the spec as it is not under our control, but why not?

Oops, I missed the sentence "controlled by the author of the application" in the spec. In an ideal world, we can control lxqt.org, but that's another long story (https://github.com/lxqt/lxqt/issues/1829)

**Reply by stefonarch** [reply]
Ecluding the 3 risky ones are remaining 
* ScreenGrab
* Qps
* QTerminal
* ObConf-qt
* lxqt-archiver
* Configuration Center (lxqt-config)
* Qlipper? This is still hanging: https://github.com/lxqt/lxqt/discussions/2372#discussioncomment-5096922

**Reply by yan12125** [reply]
* `qterminal.desktop` is used by libfm-qt: https://github.com/lxqt/libfm-qt/blob/1.3.0/data/terminals.list#L58
* `lxqt-config.desktop` is used in menu files: https://github.com/search?q=repo%3Alxqt%2Flxqt-menu-data%20lxqt-config.desktop&type=code. Users may do some customizations based on existing menu files. If so, they need to change `lxqt-config.desktop` in their customized files.

Apparently `*.desktop` filenames are used in many places. Instead of investigating which are safe to change and which are not, I prefer to keep current names and use "reverse DNS" names for new desktop files only.

**Comment by yan12125** [comment]
> The name of the desktop entry **should** follow the "reverse DNS" convention

In IETF RFC documents, "SHOULD" has the same meaning as "RECOMMENDED" :)

https://datatracker.ietf.org/doc/html/rfc2119

**Comment by stefonarch** [comment]
So if any remain LXimage-qt, ScreenGrab, Qps, lxqt-archiver  and ObConf-qt  on the quite safe side which we could port to the "recommended" standard.

**Reply by yan12125** [reply]
lximage-qt and lxqt-archiver are not "safe" as they involve file associations as mentioned above. Others might be safe.

**Reply by stefonarch** [reply]
Oops, right, too many edges ;) Going to close the PR.
